kotlinx-coroutines-core/common/src/internal/SegmentQueueSynchronizer.kt

```diff
@@ -0,0 +1,606 @@
+     * to the first cell and puts it value into it. After that, this [suspend] should come and
+     * grab the value. However, if the [synchronous][SYNC] resumption mode is used, the concurrent
+     * [resume] can mark its cell as [broken][BROKEN]; thus, this [suspend] invocation comes to the
+     * broken cell and fails. The typical patter is retrying both operations, the one that
+     * failed on [suspend] and the one that failed on [resume], from the beginning.
+     */
+    @Suppress("UNCHECKED_CAST")
+    fun suspend(cont: Continuation<T>): Boolean {
+        // Increment `enqIdx` and find the segment
+        // with the corresponding id. It is guaranteed
+        // that this segment is not removed since at
+        // least the cell for this [suspend] invocation
+        // is not in the `CANCELLED` state.
+        val curTail = this.tail.value
+        val enqIdx = enqIdx.getAndIncrement()
+        val segment = this.tail.findSegmentAndMoveForward(id = enqIdx / SEGMENT_SIZE, startFrom = curTail,
+            createNewSegment = ::createSegment).segment
+        assert { segment.id == enqIdx / SEGMENT_SIZE }
+        // Try to install the continuation in the cell,
+        // this is the regular path.
+        val i = (enqIdx % SEGMENT_SIZE).toInt()
+        if (segment.cas(i, null, cont)) {
+            // The continuation is successfully installed,
+            // add a cancellation handler if it is cancellable
+            // and complete successfully.
+            if (cont is CancellableContinuation<*>) {
+                cont.invokeOnCancellation(SQSCancellationHandler(segment, i).asHandler)
+            }
+            return true
+        }
+        // The continuation installation failed. This can happen only
+        // if a concurrent `resume` comes earlier to this cell and put
+        // its value into it. Note, that in `SYNC` resumption mode
+        // this concurrent `resume` can mark the cell as broken.
+        //
+        // Try to grab the value if the cell is not in the `BROKEN` state.
+        val value = segment.get(i)
+        if (value !== BROKEN && segment.cas(i, value, TAKEN)) {
+            // The elimination is successfully performed,
+            // resume the continuation with the value and complete.
+            cont.resume(value as T)
+            return true
+        }
+        // The cell is broken, this can happen only in `SYNC` resumption mode.
+        assert { resumeMode == SYNC && segment.get(i) === BROKEN }
+        return false
+    }
+
+    /**
+     * Tries to resume the next waiter and returns `true` if
+     * the resumption succeeds. However, it can fail due to
+     * several reasons. First of all, if the [resumption mode][resumeMode]
+     * is [synchronous][SYNC], this [resume] invocation may come
+     * before [suspend] and mark the cell as [broken][BROKEN];
+     * `false` is returned in this case. At the same time, according
```

Comment:
"At the same time" -> "Similarly"? Though I don't see much similarity here. "Another reason for [resume] to fail is..."?

Reply:
"according to X" commonly means "if we are to believe X", so here it may be a better idea to just say "if the cancellation is in simple mode and the next waiter is cancelled".